imap/jmap_backup.c

```diff
@@ -1548,22 +1562,19 @@ static int restore_message_list_cb(const mbentry_t *mbentry, void *rock)
                    record->uid, msgid ? msgid : "");
         }
 
-        /* Remove $restored flag from message */
+        /* Add message to plan for mailbox */
         if (!(rrock->jrestore->mode & DRY_RUN) && userflag >= 0
             && (record->user_flags[userflag/32] & (1<<userflag%31))) {
-            syslog(log_level,
-                   "UID %u: removing $restored flag (%d)", record->uid, userflag);
-            struct index_record newrecord;
-            /* copy the existing index_record */
-            memcpy(&newrecord, record, sizeof(struct index_record));
-            newrecord.user_flags[userflag/32] &= ~(1<<userflag%31);
-            r = mailbox_rewrite_index_record(mailbox, &newrecord);
-            if (r) {
-                syslog(LOG_ERR,
-                       "IOERROR: failed to rewrite index record for %s:%u",
-                       mbentry->name, record->uid);
-                return 0;
+            struct mailbox_plan *plan =
+                hash_lookup(mailbox->name, mrock->mailboxes);
+            if (!plan) {
+                /* Create a plan for this mailbox */
+                plan = xzmalloc(sizeof(struct mailbox_plan));
+                hash_insert(mailbox->name, plan, mrock->mailboxes);
             }
+
+            /* Add this msgno to the unflag array */
+            arrayu64_append(&plan->unflag, record->recno);
```

Comment:
Unfortunately recno isn't stable over an unlock :(  - if the mailbox repacks, the recno will be wrong - so we're going to have to record them by UID instead.